packages/embark-ipfs/src/index.js

```diff
@@ -16,51 +14,106 @@ class IPFS {
     this.fs = embark.fs;
     this.storageProcessesLauncher = null;
-    this.registered = false;
+    this._config = null; // backing variable for config property
+    this._enabled = null; // backing variable for enabled property
 
-    this.events.request("runcode:whitelist", 'ipfs-api', () => { });
-    this.events.request("runcode:whitelist", 'embarkjs', () => { });
-    this.events.request("runcode:whitelist", 'embarkjs-ipfs', () => { });
+    if (this.enabled && this.config === {}) {
+      console.warn('\n===== IPFS module will not be loaded =====');
+      console.warn(`IPFS is enabled in the config, however the config is not setup to provide a URL for IPFS and therefore the IPFS module will not be loaded. Please either change the ${'config/storage > upload'.bold} setting to IPFS or add the IPFS config to the ${'config/storage > dappConnection'.bold} array. Please see ${'https://embark.status.im/docs/storage_configuration.html'.underline} for more information.\n`);
+      return;
+    }
+    if (!this.enabled) {
+      this.embark.registerConsoleCommand({
+        matches: cmd => cmd === "ipfs" || cmd.indexOf('ipfs ') === 0,
+        process: (_cmd, cb) => {
+          console.warn(__(`IPFS is disabled or not configured. Please see %s for more information.`, 'https://embark.status.im/docs/storage_configuration.html'.underline));
+          cb();
+        }
+      });
+      return;
+    }
+
+    this.events.request("runcode:whitelist", 'ipfs-api', () => {});
+    this.events.request("runcode:whitelist", 'embarkjs', () => {});
+    this.events.request("runcode:whitelist", 'embarkjs-ipfs', () => {});
     this.events.on("storage:started", this.registerIpfsObject.bind(this));
     this.events.on("storage:started", this.connectEmbarkJSProvider.bind(this));
 
     let plugin = options.plugins.createPlugin('ipfsplugin', {});
     plugin.registerActionForEvent("pipeline:generateAll:before", this.addEmbarkJSIpfsArtifact.bind(this));
 
     this.events.request("storage:node:register", "ipfs", (readyCb) => {
-      console.dir("--- ipfs readyCb")
-      console.dir('--- registering ipfs node')
       this.events.request('processes:register', 'storage', {
         launchFn: (cb) => {
           this.startProcess(cb);
         },
-        stopFn: (cb) => { this.stopProcess(cb); }
+        stopFn: (cb) => {
+          this.stopProcess(cb);
+        }
       });
       this.events.request("processes:launch", "storage", (err) => {
-        readyCb()
+        readyCb(err);
       });
-      this.registerServiceCheck()
+      this.registerServiceCheck();
     });
 
     this.events.request("storage:upload:register", "ipfs", (readyCb) => {
       let upload_ipfs = new UploadIPFS({
         buildDir: this.buildDir || 'dist/',
-        storageConfig: this.config.storageConfig,
-        configIpfsBin: this.config.storageConfig.ipfs_bin || "ipfs",
+        storageConfig: this.storageConfig,
+        configIpfsBin: this.storageConfig.ipfs_bin || "ipfs",
```

Comment:
It's usually better to use `config.storageConfig`, because if the configs changes, example from different tests changing it using requests, this property here would not be changed. 
Another way to make it more obvious would be to use a request to get the storage config before creating the object

Reply:
`this.config` should not change until `EmbarkIPFS` is reinstantiated, unless the tests are manipulating the `.config` property of this instantiated object directly (which is probably not a good thing anyway). Could you please point to a place where tests modify the `.config` property at run time?

The purpose of changing this was because I've added a calculated property called `config` that extracts the config needed based on Embark's context. IOW, if we are in an upload context, `this.config` will be the storage config's `upload` property. If we are in a run context, `this.config` will be the storage config's `dappConnection` property. In the code, we were using upload URLs in run contexts and it provided the basis for logging errors (such as service checks using an upload URL for IPFS when swarm was the storage provider for the DApp).

Reply:
Here: https://github.com/embark-framework/embark/blob/refactor_5_0_0/packages/embark-test-runner/src/test.js#L193

I check the test config (in `config()`) and if the storage config is different, I call a request for the storage config to be modified.

It does change all the modules' `this.config` objects, since they all point to the same instance. I agree that it's not the most intuitive and a better approach would be to use a request to get the latest config whenever we want to set up something using said config.

Reply:
Shouldn't that still work then in this case because `this.storageConfig = embark.config.storageConfig` is set in the constructor? (https://github.com/embark-framework/embark/pull/1783/files#diff-74119f707fe4eb584e5816805f14d579R17).

The issue might be that `this.config` is calculated once and never changed. In that case, I should remove the backing variable so the property is always calculated instead of calculated once only.

Reply:
> Shouldn't that still work then in this case because this.storageConfig = embark.config.storageConfig

No, because the `storageConfig` of `config` can be overridden, so it doesn't point to the same object.
Only `config` the same as far as I know

Reply:
I dug in to this more, and it feel like this needs to be changed. The tests are forcing us to change the way we are running our modules, which should be a red flag. The way tests look like they want to work right now is:
1. run storage module (include spin up process)
2. test overrides config
3. sends a request to restart storage module
4. storage module handles request to restart, stops the storage process, gets updated config, then restarts the process and module using the new config.

IMO, the tests should run exactly the same as Embark being run from the start.  Meaning,
1. In run context, embark starts the storage module using `engine.registerModuleGroup("storage");` which passes a config to the constructor of `embark-storage`
2. In test context, embark starts the storage module using `engine.registerModuleGroup("storage");` which passes a config to the constructor of `embark-storage`

And the test flow would be, at each configuration update in each test:
1. Kill the storage module if started
2. Update the storage config based on the configuration used in the test.
3. Start the storage module in the exact same way embark does, by registering a module package, which will pass the updated config.

I would really like to discuss this more broadly with the team if possibly.

Reply:
Those are good points. I wouldn't mind discussing with the team either